design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,341 @@
+use cast::{get_provider, get_account};
+use cast::starknet_commands::{declare, deploy, invoke, call};
+(...)
+
+fn make_account(provider: &JsonRpcClient<HttpTransport>) -> Result<SingleOwnerAccount<&'a JsonRpcClient<HttpTransport>, LocalWallet>> {
+  let mut prefunder = get_account("user", &provider)?; // the user that will prefund new account 
+  let user = create_account("user1", &provider);
+  let prefund_account = invoke("0x123", "deposit", [1234, user.address], &prefunder);
+  deploy_account("user1", &provider);
+  get_account("user1", &provider)?
+}
+
+fn main() {
+  ler provider = get_provider("http://127.0.0.1:5050/rpc")?;
+  let user = make_account(provider);
+
+  let declared_contract = declare("mycontract", &user, "/path/to/Scarb.toml");
+  let contract = deploy(&declared_contract.class_hash, [], &user);
+
+  let dispatcher = DispatcherMyContract { contract.contract_address };
+
+  skip_if_done(|| {
+        dispatcher.increase_balance(5);
+    });
+  let get_balance = skip_if_done(|| {
+        dispatcher.get_balance();
+    });
+
+  let called_balance = call(&contract.contract_address, "get_balance", [], "latest"); 
+}
+```
+
+### example state file
+The state file by default should be written to the current working directory, with a name `<script file name>.state.json`.
+Its schema could look like this:
+
+```json
+{
+  "version": 1,
+  "transactions" : {
+    "create_account": {
+      "arguments": {
+        "name": "whatever",
+        (...)
+      },
+      "output": {
+        "address": "0x123",
+        "max_fee": "0x321"
+      },
+      "status": "accepted",
+      "timestamp": (...)
+    },
+    "prefund_account": {
+      "arguments": {
+        (...)
+      },
+      "output": {
+        "error": "..."
+      },
+      "status": "rejected",
+      "timestamp": (...)
+  }
+}
+```
+
+Having this, we could allow users to `--retries` or `--verify` transactions if something unexpected happens,
+some transactions were rejected because of faulty script, too low user eth balance etc.
+
+If we were to implement dry run mode that runs by default, the state should only be written to a file when we pass `--broadcast`
+flag; otherwise the output should be directed to stdout.
+
+The file should be used to see, if given action/transaction was already taken - it should take under account the function
+name as well as its parameters, to distinguish functions from one another (eg. user creates two accounts, we should check
+state file records, one by one, and check if a function `account new` was already invoked with specified parameter `username` 
+(and others, if supplied)).
+
+### Miscellaneous
+
+There would be a new subcommand `script` that would be invoked with required parameter - a path to script file, with
+`*.s.cairo` extension. Proposed interface:
+
+- `--gas-multiplier` - relative percentage by which to multiply add gas estimates (perhaps passed directly to sncast)
```

Comment:
Should we somehow allow explicitly specifying the maximum gas for the transaction in code? Perhaps this could be done with `#[...]` annotation so at least it is limited for a specific function. This way the user could split their code into functions with specific limits.

Though an ideal solution would probably be limiting it per function call (but how exactly? and another argument or some other way of configuring it?).

Reply:
This would be a goal for later iterations - basically I think we should allow setting max_fee per function, so it would probably be another argument to the function. I view this as more of a implementation issue though, that could be solved while implementing it.